Remove build dependency on PCL reference assemblies

mellinoe opened the pull request:
A variety of small changes were needed to allow this:

* MSBuild really would prefer if you targeted a well-known profile. There's not really a straightforward way to opt-out of the whole process that I'm aware of. The changes in dir.props and dir.targets are the set of hacks I found that @ericstj  and I found that allow us to use a dummy target framework profile without needing to have any reference assemblies installed on the box. If there is a cleaner or easier way to do this, let me know. This is the ugliest and hackiest part of this change.

* Move all of the projects off of explicit portable profiles. Our build already ignores most of the things in PCL profile targeting packs unless you explicitly opt-in to them. A few of our projects were doing this: System.Collections.Immutable, System.Composition, and System.Threading.Tasks.Dataflow.

* Remove unused usings. Some libraries had using statements for namespaces that were not actually contained in any of the assemblies that they were referencing via NuGet. This was apparently legal if those namespaces WERE contained in some assembly in your profile's targeting pack, but generated build errors if you were "not targeting" a profile.

* Add an mscorlib facade for libraries which use WinRT. Windows.winmd includes a reference to mscorlib for all core types (System.Object, etc.), so a facade that forwards from mscorlib -> System.Runtime is needed. Previously, this was being picked up from the PCL profile by default (and was not disabled by us). I've added a project for this special mscorlib "Design-Time Facade", and referenced it from .NET Native libraries which use WinRT. We should place this facade into a NuGet package and reference that from these projects in the future.

We will probably need to wait until after Chris's packaging work (#6047) is completed to avoid unnecessarily delaying it with merge conflicts from this.

@ericstj , @weshaggard

mellinoe commented on dir.props:
```diff
@@ -412,4 +412,14 @@
+  <PropertyGroup>
+    <!-- We don't use any of MSBuild's resolution logic for resolving the framework, so just set these two properties to any folder that exists to skip
+         the GenerateReferenceAssemblyPaths task (not target) and to prevent it from outputting a warning (MSB3644). -->
+    <_TargetFrameworkDirectories>$(MSBuildThisFileDirectory)/Documentation</_TargetFrameworkDirectories>
```
I suspect we will want to change this :smile: . I believe this does not work if it is set to the root of corefx (i.e. `$(MSBuildThisFileDirectory)`) because the `bin` directory is underneath that and seemed to be confusing the MSBuild targets as to which files to copy into output directories. I believe `Tools` will also be problematic for the same reason (most of our framework libraries end up living under `Tools` as well).

mellinoe commented:
Addresses #5901

jasonwilliams200OK commented:
<img src=https://camo.githubusercontent.com/de05120e7fb2f0b53736c3b4873c671c81109c74/687474703a2f2f692e696d6775722e636f6d2f76774d696e2e676966 />

Sorry couldn't help myself :smile:

stephentoub commented:
@dotnet-bot test this please

krytarowski commented:
NetBSD port is waiting for it.

mellinoe commented:
@dotnet-bot Test Innerloop Ubuntu Release Build and Test

#6153

mellinoe commented:
@dotnet-bot Test Innerloop OSX Debug Build and Test
Test Innerloop Ubuntu Debug Build and Test

mellinoe commented:
Rebased this against master again. Currently running another build to validate the changes again, but this should be ready to go in momentarily.

mellinoe commented:
@dotnet-bot Test Innerloop Windows_NT Debug Build and Test 
Test Innerloop Windows_NT Release Build and Test

mellinoe commented:
@dotnet-bot Test Innerloop OSX Release Build and Test

(https://github.com/dotnet/corefx/issues/6321)

mellinoe commented:
@weshaggard @ericstj Could you guys take a quick look at the build-level changes and see if they look reasonable? The rest of the changes are just random changes to deal with the fallout from not using the PCL reference assemblies.

mellinoe commented:
@dotnet-bot  Test Innerloop OSX Release Build and Test

(Tasks tests that have already been fixed)

mellinoe commented:
@dotnet-bot Test Innerloop Windows_NT Debug Build and Test

mellinoe commented:
Re-opening this to see if the Windows debug build wants to start this time...

krytarowski commented:
I don't know what's going on with this task... but I switched to porting LLILC to NetBSD. Once it will be merged, I will be back here.

jasonwilliams200OK commented:
@krytarowski, you can't build managed code unless CLI is ported to NetBSD: https://github.com/dotnet/cli/issues/1532.

krytarowski commented:
LLILC built for NetBSD... patches are pending.

mellinoe commented:
This has been ready to go in for a few days, but the CI system has been having issues since then, and I haven't gotten a clean build all at once (all of the individual jobs have passed at one point or another at this point, though).

mellinoe commented:
Update: This actually causes some issues in our internal build when mirrored in. I need to investigate those before this goes in

weshaggard commented:
LGTM beyond the few comments.

mellinoe commented:
I updated the build targets in our internal build that were causing issues with thihs change, so that part should no longer be an issue.

mellinoe commented:
@weshaggard Could you take a look at the latest commit? I simplified the MSBuild overrides and added comments for them. They are now only in dirs.targets rather than both places.

mellinoe commented:
@dotnet-bot Test this please

weshaggard commented:
LGTM

mellinoe commented:
Spoke with @weshaggard offline, going to merge this in now and add another small comment to the msbuild overrides as a follow-up.

krytarowski commented:
Thank you!
